xformers/components/attention/csrc/cuda/attention.cu

```diff
@@ -0,0 +1,592 @@
+#pragma unroll
+  for (int64_t q_item_idx = 0; q_item_idx < kBlockSizeQ; q_item_idx++) {
+    s_prime[q_item_idx] = s_prime[q_item_idx] * m_delta[q_item_idx];
+#pragma unroll
+    for (int64_t k_item_idx = 0; k_item_idx < kBlockSizeK; k_item_idx++)
+      s_prime[q_item_idx] += s_delta[q_item_idx][k_item_idx];
+
+    m_prime[q_item_idx] = m_i[q_item_idx];
+  }
+}
+
+template <
+    typename scalar_t,
+    typename vec_t,
+    int kBlockSizeK,
+    int kBlockSizeQ,
+    int BUFFER_SIZE>
+__device__ void compute_loop(
+    vec_t* query_block[kBlockSizeQ],
+    vec_t* key_i,
+    vec_t* value_i,
+    scalar_t m_prime[kBlockSizeQ],
+    scalar_t s_prime[kBlockSizeQ],
+    vec_t buffer[kBlockSizeQ][BUFFER_SIZE] /*TODO fix me*/,
+    int64_t K) {
+  scalar_t si[kBlockSizeQ][kBlockSizeK] = {0};
+  compute_dot<scalar_t, vec_t, kBlockSizeK, kBlockSizeQ>(
+      query_block, key_i, si, K);
+
+  scalar_t m_i[kBlockSizeQ];
+  compute_max<scalar_t, kBlockSizeK, kBlockSizeQ>(si, m_prime, m_i);
+
+  scalar_t m_delta[kBlockSizeQ];
+  scalar_t s_delta[kBlockSizeQ][kBlockSizeK];
+
+  compute_scaling_coeffs<scalar_t, kBlockSizeK, kBlockSizeQ>(
+      m_i, m_prime, si, m_delta, s_delta);
+
+  compute_final_mult<scalar_t, vec_t, kBlockSizeK, kBlockSizeQ, BUFFER_SIZE>(
+      value_i, s_delta, m_delta, buffer, K);
+
+  update_scaling_coeffs<scalar_t, kBlockSizeK, kBlockSizeQ>(
+      m_delta, m_i, s_delta, m_prime, s_prime);
+}
+
+template <
+    typename scalar_t,
+    typename vec_t,
+    int kBlockSizeQ,
+    int WARP_SIZE,
+    int BUFFER_SIZE>
+__device__ __forceinline__ void aggregate_coeffs(
+    scalar_t m_prime[kBlockSizeQ],
+    scalar_t s_prime[kBlockSizeQ],
+    vec_t buffer[kBlockSizeQ][BUFFER_SIZE] /*TODO fix me*/,
+    int64_t K) {
+  constexpr int kVecSize = sizeof(vec_t) / sizeof(scalar_t);
+  for (int64_t q_item_idx = 0; q_item_idx < kBlockSizeQ; q_item_idx++) {
+    scalar_t m_i = m_prime[q_item_idx];
+    scalar_t s_i = s_prime[q_item_idx];
+    m_prime[q_item_idx] = warpMax<scalar_t, WARP_SIZE>(m_prime[q_item_idx]);
+    scalar_t m_delta = std::exp(m_i - m_prime[q_item_idx]);
+    scalar_t s_delta = s_i * m_delta;
+    s_delta = warpSum<scalar_t, WARP_SIZE>(s_delta);
+    s_prime[q_item_idx] = s_delta;
+    for (int64_t k = 0; k < K / kVecSize; k += 1) {
+      vec_t tmp = buffer[q_item_idx][k];
+      iMul<scalar_t>(m_delta, &tmp);
+      tmp = warpSum<vec_t, WARP_SIZE>(tmp);
+      buffer[q_item_idx][k] = tmp;
+    }
+  }
+}
+
+template <
+    bool first,
+    typename scalar_t,
+    typename vec_t,
+    int kBlockSizeK,
+    int kBlockSizeQ,
+    int BUFFER_SIZE,
+    int WARP_SIZE>
+struct UnrollLoop {
+  static __device__ __forceinline__ void eval(
+      vec_t* query_block[kBlockSizeQ],
+      at::TensorAccessor<scalar_t, 2> key,
+      at::TensorAccessor<scalar_t, 2> value,
+      scalar_t m_prime[kBlockSizeQ],
+      scalar_t s_prime[kBlockSizeQ],
+      vec_t buffer[kBlockSizeQ][BUFFER_SIZE] /*TODO fix me*/,
```

Comment:
The `TODO: fix me` is not resolved yet. It consists of the limitations that I pointed in the end of the PR description, regarding limits on the size of `K` (the dimension of the features).

Making this completely generic (without just increasing `BUFFER_SIZE`) is probably possible, although I haven't spent much time thinking about it for now.

Reply:
understood, I would just elaborate in one of the comments so that either somebody else or the future you has more context when stumbling upon this fixme ?